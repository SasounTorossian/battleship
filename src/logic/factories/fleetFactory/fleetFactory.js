const fleetFactory = () => {

    // Takes segment and marks it as hit.
    function hit(segment) { this.position[segment] = "X" }

    // Checks all segments in position to see if ship is fully sunk.
    function isSunk() { return !!this.position.length && this.position.every(segment => segment === "X") }

    // Checks if all ships in the fleet are sunk.
    const checkIfAllShipsSunk = () => { return ships.every(ship => ship.isSunk()) }

    //TODO: Change "type" to "class" for better accuracy.
    const ships = [
        {
            type: "destroyer",
            length: 2,
            position: [],
            orientation: [
                [0, 1],
                [0, 10]
            ],
            hit: hit,
            isSunk: isSunk,
        },
        {
            type: "submarine",
            length: 3,
            position: [],
            orientation: [
                [0, 1, 2],
                [0, 10, 20]
            ],
            hit: hit,
            isSunk: isSunk,
        },
        {
            type: "cruiser",
            length: 3,
            position: [],
            orientation: [
                [0, 1, 2],
                [0, 10, 20]
            ],
            hit: hit,
            isSunk: isSunk,
        },
        {
            type: "battleship",
            length: 4,
            position: [],
            orientation: [
                [0, 1, 2, 3],
                [0, 10, 20, 30]
            ],
            hit: hit,
            isSunk: isSunk,
        },
        {
            type: "carrier",
            length: 5,
            position: [],
            orientation: [
                [0, 1, 2, 3, 4],
                [0, 10, 20, 30, 40]
            ],
            hit: hit,
            isSunk: isSunk,
        },
    ]

    return {
        ships,
        checkIfAllShipsSunk,
    }

} 

export default fleetFactory