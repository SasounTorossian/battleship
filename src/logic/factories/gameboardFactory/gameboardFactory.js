const gameboardFactory = () => {

    let gameboard = []

    // Initialize gameboard as 1D array of 100 objects.
    const initGameboard = () => {
        for(let i=0; i<100; i++) {
            let gameSquare = { 
                id: i, 
                occupied: false, 
                hit: false,
                ship: {}
            }
            gameboard.push(gameSquare)
        }

        return gameboard
    }

    // Handles oncoming clicks on the gameboard.
    const clickHandler = (e, squareID) => {
        console.log(e.target);
        console.log(squareID);
    }

    // Displays gameboard 1D array as 2D array of hits. 
    // Used for debugging with console.table().
    const displayArrayHit = () => {
        let displayArray = []
        let displayGameboard = gameboard.map(el => +el["hit"])

        while(displayGameboard.length) {
            displayArray.push( displayGameboard.splice(0, 10) )
        }

        return displayArray
    }

    // Displays gameboard 1D array as 2D array of IDs. 
    // Used for debugging with console.table().
    const displayArrayId = () => {
        let displayArray = []
        let displayGameboard = gameboard.map(el => el["id"])

        while(displayGameboard.length) {
            displayArray.push( displayGameboard.splice(0, 10) )
        }

        return displayArray
    }

    // Displays gameboard 1D array as 2D array of occupied spaces. 
    // Used for debugging with console.table().
    const displayArrayOccupied = () => {
        let displayArray = []
        let displayGameboard = gameboard.map(el => +el["occupied"])

        while(displayGameboard.length) {
            displayArray.push( displayGameboard.splice(0, 10) )
        }

        return displayArray
    }

    // Displays gameboard 1D array as 2D array of objects. 
    // Used for debugging with console.table().
    const displayArray = () => {
        let displayArray = []

        while(gameboard.length) {
            displayArray.push( gameboard.splice(0, 10) )
        }

        return displayArray
    }

    // Check if ship placement interferes with existing ships.
    const checkShipCollision = (position, orientation) => {
        return orientation.some(index => gameboard[position + index].occupied === true )
    }

    // Check if ship is out of bounds on horizontal axis.
    const checkHorizontalOutOfBounds = (position, orientation) => {
        return orientation.some(index => (position % 10) + index >= 10)
    }

    // Check if ship is out of bounds on vertical axis.
    const checkVerticalOutOfBounds = (position, orientation) => {
        return orientation.some(index => (position + index >= 100))
    }

    // Places all fleet ships onto gameboard
    const placeAllShips = (fleet) => {
        fleet.forEach(ship => placeShip(ship))
    }

    // Places each ship object on the gameboard according to its parameters.
    const placeShip = (ship, testPosition = false, testOrientation = false) => {

        // Randomly select (or test select) horizontal or vertical orientation of ship.
        let randomOrientation
        let currentShipOrientation
        if(testOrientation !== false) { randomOrientation = testOrientation }
        else { randomOrientation = Math.floor(Math.random() * ship.orientation.length) }
        currentShipOrientation = ship.orientation[randomOrientation]
        
        // Randomly select (or test select) starting positition of ship.  
        let randomPosition
        if(testPosition !== false) { randomPosition = testPosition }
        else { randomPosition = Math.floor(Math.random() * gameboard.length) }

        if((randomOrientation ? 
            checkVerticalOutOfBounds(randomPosition, currentShipOrientation) :
            checkHorizontalOutOfBounds(randomPosition, currentShipOrientation)) || 
            checkShipCollision(randomPosition, currentShipOrientation))  
        { 
            placeShip(ship) // recursively call function to properly place ship.
        }
        else {
            currentShipOrientation.forEach(index => gameboard[randomPosition + index].occupied = true) // Populate gameboard.
            currentShipOrientation.forEach(index => gameboard[randomPosition + index].ship = ship) // Populate gameboard.
            currentShipOrientation.forEach(index => ship.position.push(randomPosition + index)) // Populates position variable in ship.
        }
    }

    // Find which ship is occupying a particular position. 
    const findShipFromPosition = (ships, pos) => {
        return ships.find(ship => {
            return ship.position.indexOf(pos) !== -1
        })
    }

    // TODO: Could possible move this to hit() function in ship objects.
    // Find which ship index corresponds to particular position. 
    const findIndexFromPosition = (ship, pos) => {
        return ship.position.indexOf(pos)
    }

    // TODO: Need to handle hit and set gameboard[hitPosition].hit = true even if no ship. Terrible Logic.
    // Takes gameboard square as an ID, and passes hit to ship if it exists.
    const receiveAttack = (ships, hitPosition) => {
        let ship
        let index

        if(ships !== undefined && hitPosition !== undefined) {
            ship = findShipFromPosition(ships, hitPosition)
        }
        else { return }

        if(ship !== undefined) {
            index = findIndexFromPosition(ship, hitPosition)
        }
        else { return }

        gameboard[hitPosition].hit = true
        ship.hit(index)
    }

    return {
        gameboard,
        initGameboard,
        clickHandler,
        displayArrayHit,
        displayArrayId,
        displayArrayOccupied,
        displayArray,
        placeAllShips,
        placeShip,
        receiveAttack
    }
}

export default gameboardFactory