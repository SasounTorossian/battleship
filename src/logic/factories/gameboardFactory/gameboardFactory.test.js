import gameboardFactory from "./gameboardFactory"
import fleetFactory from "../fleetFactory/fleetFactory"

test("test gameboardFactory is valid", () => {
    expect(gameboardFactory).toBeTruthy()
})

test("test gameboardFactory starts with empty gameboard", () => {
    const newGameboard = gameboardFactory()
    expect(newGameboard.gameboard).toStrictEqual([])
})

test("test initGameboard creates full array", () => {
    let testArray = []
    for(let i=0; i<100; i++) {
        let gameSquare = {
            id: i, 
            occupied: false,
            hit: false,
            ship: {}
        }
        testArray.push(gameSquare)
    }

    const newGameboard = gameboardFactory()
    expect(newGameboard.initGameboard()).toStrictEqual(testArray)
})

test('test ID array formatting console.table() debugging', () => {
    let testArray = [
        [0,   1,  2,  3,  4,  5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
        [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
        [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
        [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
        [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
        [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
        [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]
    ]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    expect(newGameboard.displayArrayId()).toStrictEqual(testArray)
})

test('test occupied array formatting console.table() debugging', () => {
    let testArray = [
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    expect(newGameboard.displayArrayOccupied()).toStrictEqual(testArray)
})

test('test horizontal destroyer placement in correct area', () => {
    let testArray = [
        [1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]

    const newFleet = fleetFactory()
    const destroyer = newFleet.ships[0]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    const placeShipSpy = jest.spyOn(newGameboard, "placeShip")
    placeShipSpy.mockImplementation(() => testArray)

    newGameboard.placeShip(destroyer)

    expect(newGameboard.displayArrayOccupied()).toMatchSnapshot()
})

test('test vertical destroyer placement in correct area', () => {
    let testArray = [
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]

    const newFleet = fleetFactory()
    const destroyer = newFleet.ships[0]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    const placeShipSpy = jest.spyOn(newGameboard, "placeShip")
    placeShipSpy.mockImplementation(() => testArray)

    newGameboard.placeShip(destroyer)

    expect(newGameboard.displayArrayOccupied()).toMatchSnapshot()
})

test('test horizontal destroyer placement in incorrect area', () => {
    let testArray = [
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]

    const newFleet = fleetFactory()
    const destroyer = newFleet.ships[0]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    const placeShipSpy = jest.spyOn(newGameboard, "placeShip")
    placeShipSpy.mockImplementation(() => testArray)

    newGameboard.placeShip(destroyer)

    expect(newGameboard.displayArrayOccupied()).not.toStrictEqual(testArray)
})

test('test horizontal destroyer placement in array out of bounds', () => {
    let testArray = [
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
    ]

    const newFleet = fleetFactory()
    const destroyer = newFleet.ships[0]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    const placeShipSpy = jest.spyOn(newGameboard, "placeShip")
    placeShipSpy.mockImplementation(() => testArray)

    newGameboard.placeShip(destroyer)

    expect(newGameboard.displayArrayOccupied()).not.toStrictEqual(testArray)
})


test('test vertical destroyer placement in incorrect area and array out of array bounds', () => {
    let testArray = [
        [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
    ]

    const newFleet = fleetFactory()
    const destroyer = newFleet.ships[0]

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()

    const placeShipSpy = jest.spyOn(newGameboard, "placeShip")
    placeShipSpy.mockImplementation(() => testArray)

    newGameboard.placeShip(destroyer)

    expect(newGameboard.displayArrayOccupied()).not.toStrictEqual(testArray)
})

test('test placing all ships on gameboard with no collisions', () => {
    let testArray = [
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]
    
    const newFleet = fleetFactory()
    const ships = newFleet.ships

    const newGameboard = gameboardFactory()
    newGameboard.initGameboard()
    newGameboard.placeAllShips(ships)

    let occupyArray = newGameboard.gameboard.map(square => +square.occupied)
    let totalShips = occupyArray.reduce((a, b) => a + b)

    expect(newGameboard.displayArrayOccupied()).not.toStrictEqual(testArray) // Array should not be empty after placements
    expect(totalShips).toBe(17) // If no collisions, there should be 17 spots occupied by the ships
})

// test('test carrier has position variable populated when placed vertically at position 57', () => {
//     let testArray = [
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0]
//     ]
    
//     const newFleet = fleetFactory()
//     const ships = newFleet.ships
//     let carrier = ships[4]

//     const newGameboard = gameboardFactory()
//     newGameboard.initGameboard()
//     newGameboard.placeShip(carrier, 57, 1)

//     expect(newGameboard.displayArrayOccupied()).toStrictEqual(testArray) // Array should not be empty after placements
//     expect(carrier.position).toStrictEqual([57, 67, 77, 87, 97]) // Array should not be empty after placements
// })

// test('test receiveAttack() correctly damages ship.', () => {    
//     let testHitArray = [
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
//     ]

//     const newFleet = fleetFactory()
//     const ships = newFleet.ships
//     let carrier = ships[4]

//     const newGameboard = gameboardFactory()
//     newGameboard.initGameboard()
//     newGameboard.placeShip(carrier, 57, 1) //carrier.position = [57, 67, 77, 87, 97]
//     newGameboard.receiveAttack(ships, 57)
//     newGameboard.receiveAttack(ships, 87)

//     expect(carrier.position).toStrictEqual(["X", 67, 77, "X", 97]) // Position should be updated
//     expect(newGameboard.displayArrayHit()).toStrictEqual(testHitArray) // Hit array should be updated
// }) 

// test('test receiveAttack() exits gracefully if no ship at position', () => {    
//     let testHitArray = [
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
//         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
//     ]

//     const newFleet = fleetFactory()
//     const ships = newFleet.ships
//     let carrier = ships[4]

//     const newGameboard = gameboardFactory()
//     newGameboard.initGameboard()
//     newGameboard.placeShip(carrier, 57, 1) //carrier.position = [57, 67, 77, 87, 97]
//     newGameboard.receiveAttack(ships, 98)
    
//     expect(carrier.position).toStrictEqual([57, 67, 77, 87, 97]) // Carrier should be undamaged
//     expect(newGameboard.displayArrayHit()).toStrictEqual(testHitArray) // Hit array should remain unchanged
// }) 

