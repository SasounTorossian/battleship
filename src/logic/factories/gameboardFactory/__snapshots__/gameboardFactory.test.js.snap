// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test horizontal destroyer placement in correct area 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`test vertical destroyer placement in correct area 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
]
`;
